### PR TITLE
Feat: parseJson

### LINKED ISSUE #2153: feat: JSON parsing cheatcodes
### Component

Forge

### Describe the feature you would like

The ability to easily parse JSON files within Solidity is very useful for things like reading deploy script outputs, config files, etc. Here is a proposed spec on how this should be implemented:

## Cheatcodes

The `readFile` cheatcode already exists, and we just add a `parseJson` cheatcode which takes a string of JSON and `key`, specified using the same syntax as [`jq`](https://stedolan.github.io/jq/manual/). It returns the data of each key as ABI-encoded `bytes`.

This implies `parseJson` will need to infer data types from the JSON to ABI-encode them appropriately. In particular, we need to distinguish between a hex string that's `bytes` (bytes get right-padded) vs. a hex string that's a number (numbers get left-padded). I think ethers.js has a convention for distinguishing these, we should check that convention, use the same one, document it, and make sure it's followed when the JSON output files from scripts are written. I think the convention is `0x1234` for a number and `[0x12, 0x34]` for a bytes but am not certain.

```solidity
interface Vm {
  // Reads the entire content of file to string, (path) => (data)
  function readFile(string calldata) external returns (string memory);

  // Given a string of JSON, find the provided key, (stringified json) => (ABI-encoded data)
  function parseJson(string calldata json, string calldata key) external view returns (bytes memory);
}

## forge-std

We'll also add new forge-std helpers to `Test.sol`, as shown below.

```solidity
// Everything shown here is new and not yet present in Test
contract Test {
  // Reading in deployment logs will be common, so let's include helpers for
  // them in forge-std.
  // NOTE: Not all of the below data types are correct, we need to verify them,
  // e.g. I think nonce is really a uint64.
  struct Transaction {
    uint256 txType;
    address from;
    address to;
    uint256 gas;
    uint256 value;
    bytes data;
    uint256 nonce;
  }

  struct TransactionDetail {
    bytes32 hash;
    // e.g. F0 for CREATE, this is called `type` in the output but that's a solidity keyword.
    // We should consider changing that for consistency.
    bytes1 opcode;
    string contractName;
    address contractAddress;
    Transaction transaction;
  }

  struct Receipt {
    bytes32 transactionHash;
    // --- snip, you get the idea ---
  }

  // Read in all deployments transactions.
  function readTransactions(string memory path) internal view returns (TransactionDetail[] memory) {
    string memory deployData = vm.readJson(path);
    bytes memory parsedDeployData = vm.parseJson(deployData, ".transactions[]");
    return abi.decode(parsedDeployData, (TransactionDetail[]));
  }

  // Analogous to readTransactions, but for receipts.
  function readReceipts(string memory path) internal view returns (Receipt[] memory) {
    // --- snip, you get the idea ---
  }

  // Helpers for parsing keys into types. We'd include these for all value types
  // as well as `bytes`, `string`, `uint256[]`, and `int256[]`. Only two are shown below.
  function readUint256(string memory json, string memory key) internal view returns (uint256) {
      return abi.decode(vm.parseJson(json, key), (uint256));
  }
  function readBytes32(string memory json, string memory key) internal view returns (bytes32) {
      return abi.decode(vm.parseJson(json, key), (bytes32));
  }
}
```

## Example Usage

The above would result in the following sample usage.

```solidity
contract MyTest is Test {
  string internal constant deployFile = "broadcast/Deploy.s.sol/10/run-latest.json";

  function myFunction() public {
    // Get all deployment transactions.
    TransactionDetail[] memory transactions = readTransactions(deployFile);

    // Get the name of the first contract deployed.
    string memory deployData = vm.readJson(deployFile);
    string memory contractName = abi.decode(vm.parseJson(deployData, ".transactions[0].contractName"), (string));

    // Get the nonce and transaction hash of the first contract deployed.
    uint256 nonce = readUint256(deployData, ".transactions[0].tx.nonce");
    bytes32 txHash = readBytes32(deployData, ".transactions[0].hash");
  }
}
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by odyslam**
## Motivation

I am opening a draft PR to gather feedback and improve code quality.

- fixes #2153 
- related to https://github.com/foundry-rs/forge-std/pull/138

## Checklist

- [x] Poc
- [x] Support arbitrarily nested JSON objects
- [x] Add Solidity tests
- [x] Improve error handling & code quality
- [x] Rebase master
- [ ] Merge forge-std library
- [ ] Merge docs

### CODE REVIEW

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 1]
```diff
@@ -240,6 +248,47 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+fn value_to_type(value: &Value) -> Option<ParamType> {
+    if value.is_boolean() {
+        Some(ParamType::Bool)
+    } else if value.is_string() {
+        // If string can be converted to Address, chances are that it is
+        // TODO: What about H160 that are NOT addresses?
+        match Address::from_str(value.as_str().unwrap()) {
+            Ok(addr) => Some(ParamType::Address),
+            _ => Some(ParamType::String),
+        }
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:261]: here we probably need an untagged enum and try to deserialize it in the correct hex type like this

https://github.com/foundry-rs/foundry/blob/1bfb63906c2a96112cc9292d43c95e4c4a6169ab/common/src/fmt.rs#L252-L260

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 2]
```diff
@@ -240,6 +248,47 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+        Some(ParamType::Int(256))
+    } else if value.is_array() {
+        value_to_type(&value[0])
+    } else {
+        None
+    }
+}
+fn parse_json(state: &mut Cheatcodes, json: &str, key: &str) -> Result<Bytes, Bytes> {
+    // let v: Value = serde_json::from_str(json).map_err(util::encode_error)?;
+    let values: Value = jsonpath_rust::JsonPathFinder::from_str(json, key)?.find();
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:274]: what's the benefit if `JsonPathFinder` here?

**Review by mattsse — CHANGES_REQUESTED** [review]
I think that should work

> If I understand correctly the ABI spec, structs are encoded as tuples.

exactly

for objects we should be able to convert them into `Vec<ParamType>` which will encode/decode as tuples, which would mean `value_to_type` is recursive over `Value::Object`

**Comment by mattsse** [comment]
> @mattsse What do you think of the fact that json objects are not ordered, thus it's possible that the same JSON will be serialized differently with every read of the file?

iirc serde_json::Value::Object is a sorted BTreeMap.

I think we want sorted keys, and hence sorted encoded tuples so order is consistent?

File: testdata/cheats/ParseJson.t.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        string memory path = "../testdata/fixtures/ParseJson/test.json";
+        json = cheats.readFile(path);
+    }
+
+    function test_uintArray() public {
+        bytes memory data = cheats.parseJson(json, ".uintArray");
+        uint[] memory decodedData = abi.decode(data, (uint[]));
+        assertEq(42, decodedData[0]);
+        assertEq(43, decodedData[1]);
+    }
```
[the remark this answers is not in the document]
Comment (odyslam) [reply testdata/cheats/ParseJson.t.sol:21]: > parseJson variant which does not take a second argument and reads the entire thing

Reads it into what? We have the JSON in stringified via the `readFile`. So what we could do is basically a library that reads the JSON via `readFile`, stores the string locally, and then does `parseJson(key)` on the stored inner value.

Agreed on the library for forge-std.

**Comment by odyslam** [comment]
> > @mattsse What do you think of the fact that json objects are not ordered, thus it's possible that the same JSON will be serialized differently with every read of the file?
> 
> iirc serde_json::Value::Object is a sorted BTreeMap.
> 
> I think we want sorted keys, and hence sorted encoded tuples so order is consistent?

Yup, that sounds reasonable. Just underlining it cause that affects the order the user would have to define the attributes in the struct. While they would intuitively do it via some taxonomical or tight-packing order, they would need to know they they have to list the attributes in alphabetical order in order for the tuple to be translated into the struct correctly

**Comment by mattsse** [comment]
> they would need to know they they have to list the attributes in alphabetical 

I think this is reasonable, otherwise it will cause problems if you want to parse e.g. a "transaction" json and different nodes may return a different order in the RPC response

**Comment by odyslam** [comment]
@mattsse finally made it work with arbitrarily nested objects. Switched to handling them via Tokens and it became much simpler. I would like your suggestion on:
- Better error handling, especially inside the closure to reduce `unwrap()`
- parseAddress: I tried your suggestion, but didn't make it work. I think serde follows the same logic under the hood, trying different types in-order unless one decodes successfully.
- Solidity tests pass. To run them easily, you can use this one-liner: `./target/debug/forge test --root testdata --contracts cheats -vvv --match-contract ParseJson`

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 4]
```diff
@@ -240,6 +250,95 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+/// deserialized in the same order. That means that the solidity `struct` should order it's fields
+/// alphabetically and not by efficient packing or some other taxonomy.
+fn parse_json(_state: &mut Cheatcodes, json: &str, key: &str) -> Result<Bytes, Bytes> {
+    let values: Value = jsonpath_rust::JsonPathFinder::from_str(json, key)?.find();
+    // values is an array of items. Depending on the JsonPath key, they
+    // can be many or a single item. An item can be a single value or
+    // an entire JSON object.
+    let res = values
+        .as_array()
+        .unwrap()
```
Comment (gakonst) [inline evm/src/executor/inspector/cheatcodes/ext.rs:298]: replace with `?` and `.ok_or` to convert the Option into a Result

**Review by gakonst — COMMENTED** [review]
there's a commented out test + needs a test for writeJSON, not sure i exactly get how it's supposed to be used! 

the parseJSON flow looks good

**Comment by odyslam** [comment]
@gakonst I added a new comment on the [issue](https://github.com/foundry-rs/foundry/issues/2153#issuecomment-1200389140) on a proposed API for writeJson. I haven't found something that I love, so I want to get some alignment before I get full steam ahead implementing. 

The one I have implemented right now, is only good for writing simple values (not arrays, not struct) at the first "level" of a json (so not being able to write in a nested object). I proposed a new API, but it's meh. Any idea is more than welcome.

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 5]
```diff
@@ -240,6 +250,95 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+    } else if value.is_object() {
+        let values = value
+            .as_object()
+            .unwrap()
+            .values()
+            .map(|val| value_to_token(val).unwrap())
+            .collect::<Vec<Token>>();
+        Ok(Token::Tuple(values))
+    } else {
+        Err(Token::String("Could not decode field".to_owned()))
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:284]:
```suggestion
        Err(Token::String("Could not decode field".to_string()))
```

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 6]
```diff
@@ -240,6 +250,95 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+    // values is an array of items. Depending on the JsonPath key, they
+    // can be many or a single item. An item can be a single value or
+    // an entire JSON object.
+    let res = values
+        .as_array()
+        .ok_or(util::encode_error("JsonPath did not return an array"))?
+        .iter()
+        .map(|inner| {
+            let val = inner;
+            value_to_token(val).unwrap()
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:302]: lets not unwrap here instead collect over `::Result<Vec<_>,_>`

File: evm/src/executor/inspector/cheatcodes/ext.rs [hunk 7]
```diff
@@ -240,6 +250,95 @@ fn remove_file(state: &mut Cheatcodes, path: impl AsRef<Path>) -> Result<Bytes,
...
+    let abi_encoded = abi::encode(&[Token::Bytes(abi::encode(&res))]);
+    Ok(abi_encoded.into())
+}
+
+fn write_json(
+    state: &mut Cheatcodes,
+    keys: &Vec<String>,
+    values: &Vec<String>,
+    path: impl AsRef<Path>,
+    overwrite: &bool,
```
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/ext.rs:315]: don't borrow Copy types
```suggestion
    overwrite: bool,
```

**Review by mattsse — CHANGES_REQUESTED** [review]
couple nits, otherwise ,lg

**Comment by OliverNChalk** [comment]
Hey, just testing this out locally and haven't had any issues parsing JSON files yet. I merged master in and it was fine, just one compile issue fixed here: https://github.com/proximacapital/foundry/commit/e6867e11c89e21a1dc614d2cc68dd292ec2e3537

Was wondering if there was an eta? I'll continue building out my json usage locally but struggling to see a practical (i.e. not compiling from source) way to get access to this feature in CI before it merges?

**Comment by gakonst** [comment]
`foundryup -P 2293` will get this working for you I think?

**Comment by OliverNChalk** [comment]
Build currently failing but ill def use that flag in the future, for now running off of a fork

**Review by mattsse — APPROVED** [review]
a general usage is limited due to some assumptions and limitations regarding encoding, but it is definitely useful and we should add that

needs rebase though

**Comment by mds1** [comment]
> a general usage is limited due to some assumptions and limitations regarding encoding, but it is definitely useful and we should add that

What limitations are these? 

Also @odyslam should this be taken out of draft now?

**Comment by odyslam** [comment]
Tomorrow will do the following:
- remote writeJson code (will be replaced eitherway)
- rebase
- put out of draft and ping for review. Goal to merge tomorrow

Then, I will create a new branch to start working on @mattsse suggestion for `writeJson`. 

Sorry for delaying the feature for so long.

**Comment by gakonst** [comment]
SGTM. Let's get this over the line and iterate on the `writeJSON` separately.

**Comment by mattsse** [comment]
> > a general usage is limited due to some assumptions and limitations regarding encoding, but it is definitely useful and we should add that
> 
> What limitations are these?

for structs, we currently have no way of knowing how the target struct looks like so current assumption is that the sol type matches the json object 1:1, right @odyslam?

**Comment by odyslam** [comment]
Not only that, but that also the attributes of the struct are defined in alphabetical order

**Comment by odyslam** [comment]
@mattsse will proceed with `forge-std` and docs update shortly. 

Thanks for the JSON file test fix.

**Comment by 0xPhaze** [comment]
Does this work with `bytes32` and `bytes`? I am getting some weird results there.

**Comment by OliverNChalk** [comment]
@0xPhaze I am able to parse this:
image

Into this:
image

**Comment by 0xPhaze** [comment]
Hm.. are you able to parse something like this?
image

**Comment by odyslam** [comment]
Hey, 

so currently the above will be returned as a type string. I have added a fix for that and will pr now

**Comment by odyslam** [comment]
@0xPhaze that should do it

https://github.com/foundry-rs/foundry/pull/2866

**Pull request merged**